### PR TITLE
lifecycle: integrate executor clean and package list refresh (CRAFT-32)

### PR DESCRIPTION
**Pull request by cmatsuoka**
Add step and part cleaning and package list refresh methods to the
lifecycle manager.

Signed-off-by: Claudio Matsuoka <claudio.matsuoka@canonical.com>

- [x] Have you signed the [CLA](http://www.ubuntu.com/legal/contributors/)?

-----

### CODE REVIEW

**Review by sergiusens — APPROVED** [review]

File: craft_parts/lifecycle_manager.py [hunk 1]
```diff
@@ -103,6 +103,40 @@ def project_info(self) -> ProjectInfo:
...
+
+        Cleaning a step removes its state and all artifacts generated in that
+        step and subsequent steps for the specified parts.
+
+        :para step: The step to clean.
+        :param part_names: The list of part names to clean. If not specified,
+            all parts will be cleaned and work directories will be removed.
+        """
+        if not step:
+            step = Step.PULL
```
Comment (facundobatista) [inline craft_parts/lifecycle_manager.py:119]: How weird would be to put this default in the signature above?

**Review by facundobatista — APPROVED** [review]
Approved, but see a comment, thanks!

File: craft_parts/lifecycle_manager.py [hunk 2]
```diff
@@ -103,6 +103,40 @@ def project_info(self) -> ProjectInfo:
...
+        it shouldn't be updated between planning and execution.
+
+        :param system: Also refresh the list of available build packages to
+            install on the host system.
+        """
+        packages.Repository.refresh_stage_packages_list(
+            application_name=self._application_name, target_arch=self._target_arch
+        )
+
+        if system:
```
Comment (cjp256) [inline craft_parts/lifecycle_manager.py:137]: Is it expected to be called more than once with a conditional?  Maybe they should just be separate?

File: craft_parts/lifecycle_manager.py (hunk 1 above)
Comment (cmatsuoka) [reply craft_parts/lifecycle_manager.py:119]: Good idea, will do that.

File: craft_parts/lifecycle_manager.py (hunk 2 above)
Comment (cmatsuoka) [reply craft_parts/lifecycle_manager.py:133]: We'll also need to update the overlay package list and my idea was to consolidate all three in a single call to simplify the API. Let's see how the overlay discussion goes, we can refactor later if needed.

**Pull request merged**